### PR TITLE
Allow pd-workshop sections to accept members who are in other workshops from the same organizer.

### PR DESCRIPTION
**Pull request by aoby**
Removed unique constraint from followers.[user_id, student_user_id].
Added unit tests for Section.add_student.

### CODE REVIEW

**Comment by aoby** [comment]
When asked about the `followers [user_id, student_user_id]` unique constraint,
@ashercodeorg said in slack "That constraint is both intentional and very problematic. Laurel indicated she thought the work to get rid of it was worth doing. But that there might be some subtleties (in work and teacher expectations) in getting rid of it."

All tests are passing. The functionality should remain the same in the default case with this change, and now it has a unit test. The `section.add_student` logic didn't rely on the unique constraint anyway.

**Comment by tanyaparker** [comment]
Can you give more context on the issues that triggered this change and what the impact is? What are the use cases for teachers/facilitators?

**Comment by aoby** [comment]
I happened upon the `Section.add_student` logic today while investigating a [join workshop section bug](https://github.com/code-dot-org/code-dot-org/pull/9642) and realized that it will affect workshops in an unexpected way: joining the section for a new workshop put on by the same organizer as a previous workshop you attended would remove you from the previous one. 

I chatted with @alicesteinglass, and we agreed to change this behavior for workshops.

The default logic (for classroom sections), which is unchanged, is that any student can only be in one section for a given teacher. Joining another section for the same teacher would move the student. This behavior remains unchanged, and now it has a unit test :)

I first made the change only in the model (`Section.add_student`), but the new test (`add_student with move_for_same_teacher: false does not move student`) failed due to the DB unique constraint, so I added a migration to remove that. The unique constraint is not needed for the intended default behavior (which again is still unchanged and now has a test).

**Comment by tanyaparker** [comment]
Ah, ok. So we are keeping the constraint that students can only belong in 1 section per teacher. But we allow adults to belong in multiple sections per facilitator, when we know that those sections are workshop sections. Is that correct?

**Comment by tanyaparker** [comment]
And so this fix is to remove the unique constraint so that adults can join multiple workshop sections per facilitator?

**Comment by aoby** [comment]
Organizer, not facilitator, but essentially that is correct. Every workshop has a single organizer, who is also the section owner, and may have many facilitators. Currently the facilitators have no relationship to the section.

The only code path which deviates from the existing behavior, and will _not_ move followers, is [this one](https://github.com/code-dot-org/code-dot-org/pull/9660/files#diff-2e5005fb2c3ae9819c0eb503994390feR128) from the workshop join section confirmation page, which only applies to [workshop sections](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/controllers/followers_controller.rb#L37-L40).

Though I did just realize, someone could browse directly to the workshop section join confirmation page ([/pd/workshops/join/:section_code](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/controllers/pd/workshop_enrollment_controller.rb#L81-L82)) instead of `/join/:section_code`, for a non-workshop section and run this code path. I'll add another check there to make sure it only renders for workshop sections...

**Comment by aoby** [comment]
Nevermind. If there is no workshop associated with a section code (whether or not a section exists for that code), `/pd/workshops/join/:section_code` will [render 404](https://github.com/code-dot-org/code-dot-org/blob/993a5ae59fb6698494a788ffb42ceaf3f9d68bb7/dashboard/app/controllers/pd/workshop_enrollment_controller.rb#L88-L92).

**Comment by tanyaparker** [comment]
Thanks for explaining, makes more sense while reviewing.

**Comment by aoby** [comment]
Also worth noting: the first 2 unit tests I added test the existing functionality, and succeed without any of the changes (you can test this locally on staging). 

Only the 3rd one, which validates the new functionality for workshop sections, won't work without this new functionality.

File: dashboard/test/models/section_test.rb [hunk 1]
```diff
@@ -66,4 +66,58 @@ class SectionTest < ActiveSupport::TestCase
...
+    assert_does_not_create(Follower) do
+      new_section.add_student student
+    end
+
+    # Verify student has been moved to new_section
+    refute original_section.students.exists?(student.id)
+    assert new_section.students.exists?(student.id)
+  end
+
+  test 'add_student with move_for_same_teacher: false does not move student' do
```
Comment (tanyaparker) [inline dashboard/test/models/section_test.rb:104]: I wonder if this is confusing for anyone else. Terminology wise we tend to use student and follower interchangeably (same with teacher and section owner). In this test we're confirming the follower stays in both sections, but in reality we never want this to be true for _real students_ right? Not something to block this PR, just some thoughts.

Comment (aoby) [inline dashboard/test/models/section_test.rb:104]: True. The section has a single user (i.e. owner / teacher) and multiple student_users (i.e. followers / students). Soon, sections will be allowed to have multiple users (owners) as well. The terminology is pretty confusing.

Here I was trying to stick with the existing terms. The point of this test case is merely that `section.add_student` honors the `move_for_same_teacher` bool param if it's supplied. The section doesn't really care if the user type is student or teacher. This is never validated. Teachers can (and do intentionally for workshops) join sections as followers.

We could more clearly express the intent by making them both teachers and calling one the `organizer` and the other the `attendee` or `follower`? What do you think?

Comment (tanyaparker) [inline dashboard/test/models/section_test.rb:104]: That would be clearer to me. :)

**Comment by aoby** [comment]
OK, I renamed the variables and added a comment. Please take another look.

**Comment by tanyaparker** [comment]
TY - LGTM

File: dashboard/db/migrate/20160721221335_remove_unique_constraint_from_followers_user_student_user_index.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+class RemoveUniqueConstraintFromFollowersUserStudentUserIndex < ActiveRecord::Migration
+  def up
+    remove_index :followers, [:user_id, :student_user_id]
+    add_index :followers, [:user_id, :student_user_id], unique: false
+  end
+
+  def down
+    remove_index :followers, [:user_id, :student_user_id]
+    add_index :followers, [:user_id, :student_user_id], unique: true
```
Comment (ashercodeorg) [inline dashboard/db/migrate/20160721221335_remove_unique_constraint_from_followers_user_student_user_index.rb:9]: The `down` migration is problematic, as it cannot run (successfully) after data is added that violates the uniqueness constraint. Really, the migration should probably have removed uniqueness violations (perhaps by deleting the (chronologically) oldest sections) before adding the constraint.

No action required or requested at this point.

**Pull request merged**